File: 프로그래머스/2/388353. 지게차와 크레인/지게차와 크레인.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

const int EMPTY = 0;
const int OFFSET[4][2] = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

int n, m;
vector<vector<int>> board;

bool isIn(int y, int x) { return 0 <= y && y < n && 0 <= x && x < m; }
bool isOutLine(int y, int x) { return y == 0 || y == n - 1 || x == 0 || x == m - 1; }

bool canRemove(int y, int x, int c){
    vector<vector<bool>> visited(n, vector<bool> (m, false));
    
    queue<pair<int, int>> q;
    q.push({y, x});
    visited[y][x] = true;
    
    while (!q.empty()) {
        auto [curY, curX] = q.front();
        q.pop();
        
        if (isOutLine(curY, curX)) return true;
        
        for (int dir = 0; dir < 4; dir++) {
            int nextY = curY + OFFSET[dir][0];
            int nextX = curX + OFFSET[dir][1];
            
            if (isIn(nextY, nextX) && board[nextY][nextX] == EMPTY && !visited[nextY][nextX]) {
                q.push({nextY, nextX});
                visited[nextY][nextX] = true;
            }
        }
    }
    
    return false;
}

int solution(vector<string> storage, vector<string> requests) {
    n = storage.size();
    m = storage[0].size();
    
    board = vector<vector<int>> (n, vector<int> (m));
    for (int y = 0; y < n; y++) {
        for (int x = 0; x < m; x++) {
            board[y][x] = storage[y][x];
        }
    }
    
    for (string request : requests) {
        int c = request[0];
        
        if (request.size() == 2) {      // all
            for (int y = 0; y < n; y++) {
                for (int x = 0; x < m; x++) {
                    if (board[y][x] == c) board[y][x] = EMPTY;
                }
            }
        } else {
            vector<pair<int, int>> result;
            for (int y = 0; y < n; y++) {   // outline
                for (int x = 0; x < m; x++) {
                    if (board[y][x] == c && canRemove(y, x, c)) result.push_back({y, x});
                }
            }
    
            for (auto [y, x] : result) board[y][x] = EMPTY;
        }
    }
    
    int answer = 0;
    for (int y = 0; y < n; y++) {
        for (int x = 0; x < m; x++) {
            if (board[y][x] != EMPTY) answer++;
        }
    }
        
    return answer;
}

File: LeetCode/Unknown/프로그래머스/2/388353. 지게차와 크레인/지게차와 크레인.java
import java.util.*;

class Solution {

    static final int[][] OFFSET = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
    static int n, m;

    public int solution(String[] storage, String[] requests) {
        n = storage.length;
        m = storage[0].length();

        char[][] board = new char[n][m];
        for (int i = 0; i < n; i++) {
            board[i] = storage[i].toCharArray();
        }

        for (String request : requests) {
            char c = request.charAt(0);

            // 크레인
            if (request.length() == 2) {
                for (int y = 0; y < n; y++) {
                    for (int x = 0; x < m; x++) {
                        if (board[y][x] == c) {
                            board[y][x] = '0';
                        }
                    }
                }
            }
            // 지게차
            else {
                Queue<Pos> q = new LinkedList<>();
                for (int y = 0; y < n; y++) {
                    for (int x = 0; x < m; x++) {
                        if (board[y][x] == c && canRemove(board, y, x)) {
                            q.add(new Pos(y, x));
                        }
                    }
                }

                while (!q.isEmpty()) {
                    Pos p = q.poll();
                    board[p.y][p.x] = '0';
                }
            }
        }

        int answer = 0;
        for (int y = 0; y < n; y++) {
            for (int x = 0; x < m; x++) {
                if (board[y][x] != '0') answer++;
            }
        }

        return answer;
    }

    private static boolean canRemove(char[][] board, int y, int x) {
        boolean[][] visited = new boolean[n][m];
        Queue<Pos> q = new LinkedList<>();
        q.add(new Pos(y, x));
        visited[y][x] = true;

        while (!q.isEmpty()) {
            Pos cur = q.poll();

            if (isOutLine(cur.y, cur.x)) return true;

            for (int dir = 0; dir < 4; dir++) {
                int nextY = cur.y + OFFSET[dir][0];
                int nextX = cur.x + OFFSET[dir][1];
                if (isIn(nextY, nextX) && !visited[nextY][nextX] && board[nextY][nextX] == '0') {
                    q.add(new Pos(nextY, nextX));
                    visited[nextY][nextX] = true;
                }
            }
        }

        return false;
    }

    private static boolean isOutLine(int y, int x) {
        return y == 0 || y == n - 1 || x == 0 || x == m - 1;
    }

    private static boolean isIn(int y, int x) {
        return 0 <= y && y < n && 0 <= x && x < m;
    }

    private static class Pos {
        int y, x;

        public Pos(int y, int x) {
            this.y = y;
            this.x = x;
        }
    }
}